src/app/utils/field-validation/index.ts

```diff
@@ -0,0 +1,169 @@
+ * @param {string} message Message to log
+ * @throws {Error}
+ */
+const logInvalidAnswer = (
+  formId: string,
+  formField: IField,
+  message: string,
+) => {
+  formField.fieldType
+  logger.error({
+    message: `Invalid answer: ${message}`,
+    meta: {
+      action: 'InvalidAnswer',
+      formId,
+      fieldId: String(formField._id),
+      fieldType: formField.fieldType,
+    },
+  })
+}
+
+/**
+ * Constructs a validation function from a form field. Only meant to
+ * validate single answer responses.
+ * @param formField A form field from a form object
+ */
+const constructSingleAnswerValidator = (
+  formField: IField,
+): ResponseValidator<ProcessedSingleAnswerResponse> => {
+  if (isSectionField(formField)) {
+    return constructSectionValidator()
+  } else if (isShortTextField(formField) || isLongTextField(formField)) {
+    return constructTextValidator(formField)
+  }
+  return () => left('Unsupported field type')
+}
+
+/**
+ * Single exported function that abstracts away the complexities
+ * of field validation.
+ * @param {String} formId id of form, for logging
+ * @param {Object} formField A form field from the database
+ * @param {Object} response A client-side response that is to be untrusted
+ * @throws {Error}
+ */
+export default function validateField(
+  formId: string,
+  formField: IField,
+  response: FieldResponse,
+): void {
+  if (FIELDS_TO_REJECT.includes(response.fieldType)) {
+    throw new Error(`Rejected field type "${response.fieldType}"`)
+  }
+
+  // Validate that the form field type matches the response type
+  const fieldTypeEither = isFieldTypeValid(formField, response)
+
+  if (isLeft(fieldTypeEither)) {
+    throw new Error('Invalid field type submitted')
+  }
+
+  if (isProcessedSingleAnswerResponse(response)) {
+    if (singleAnswerRequiresValidation(formField, response)) {
+      switch (formField.fieldType) {
+        /* eslint-disable no-case-declarations */
+        // Migrated validators
+        case 'section':
+        case 'textfield': // short text
+        case 'textarea': // long text
+          const validator = constructSingleAnswerValidator(formField)
+          const validEither = validator(response)
+          if (isLeft(validEither)) {
+            logInvalidAnswer(formId, formField, 'Answer not allowed')
+            throw new Error('Invalid answer submitted')
+          }
+          return
+        // Fallback for un-migrated single answer validators
+        default:
+          classBasedValidation(formId, formField, response)
+        /* eslint-enable no-case-declarations */
+      }
+    }
+  } else {
+    // fallback for processed checkbox/table/attachment responses
+    classBasedValidation(formId, formField, response)
+  }
+}
+
+/**
+ * Validate that the answers in the response adhere to the form field definition
+ * using the deprecated class-based field validators.
+ * @deprecated
+ */
+function classBasedValidation(
```

Comment:
nit: use `const` declaration for consistency

Reply:
ok